substrait/substrait-spark/src/main/scala/io/substrait/spark/expression/FunctionConverter.scala

```diff
@@ -0,0 +1,300 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.substrait.spark.expression
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.analysis.{AnsiTypeCoercion, TypeCoercion}
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.substrait.ToSubstraitType
+
+import com.google.common.collect.{ArrayListMultimap, Multimap}
+import io.substrait.`type`.Type
+import io.substrait.expression.{Expression => SExpression, ExpressionCreator, FunctionArg}
+import io.substrait.function.{ParameterizedType, SimpleExtension, ToTypeString}
+import io.substrait.utils.Util
+
+import java.{util => ju}
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters
+import scala.collection.JavaConverters.collectionAsScalaIterableConverter
+
+abstract class FunctionConverter[F <: SimpleExtension.Function, T](functions: Seq[F])
+  extends Logging {
+
+  protected val (signatures, substraitFuncKeyToSig) = init(functions)
+
+  def generateBinding(
+      sparkExp: Expression,
+      function: F,
+      arguments: Seq[FunctionArg],
+      outputType: Type): T
+  def getSigs: Seq[Sig]
+
+  private def init(
+      functions: Seq[F]): (ju.Map[Class[_], FunctionFinder[F, T]], Multimap[String, Sig]) = {
+    val alm = ArrayListMultimap.create[String, F]()
+    functions.foreach(f => alm.put(f.name().toLowerCase(ju.Locale.ROOT), f))
+
+    val calciteOperators = ArrayListMultimap.create[String, Sig]()
```

Comment:
Looks like `calciteOperators` is used for function signatures, how does it relate to operator?

Reply:
This is Sig's definition:

`case class Sig(expClass: Class[_], name: String, builder: Seq[Expression] => Expression)`

and how we create one::

```scala
  private def s[T <: Expression: ClassTag](name: String): Sig = {
    val builder = FunctionRegistryBase.build[T](name, None)._2
    Sig(scala.reflect.classTag[T].runtimeClass, name, builder)
  }
```

Please node: there is a one-to-many relationship between the name and the expression, for example, `Add`   is replaced by `DateAdd`  by `Analyzer` if we add days to date

Reply:
Thanks for your explanation!